TTS/tts/models/overflow.py

```diff
@@ -0,0 +1,325 @@
+            text_len: :math:`[B]`
+            mels: :math:`[B, T_out, C]`
+            mel_len: :math:`[B]`
+        """
+        text, text_len, mels, mel_len = self.preprocess_batch(text, text_len, mels, mel_len)
+        encoder_outputs, encoder_output_len = self.encoder(text, text_len)
+        z, z_lengths, logdet = self.decoder(mels.transpose(1, 2), mel_len)
+        log_probs, fwd_alignments, transition_vectors, means = self.neural_hmm(
+            encoder_outputs, encoder_output_len, z, z_lengths
+        )
+
+        outputs = {
+            "log_probs": log_probs + logdet,
+            "alignments": fwd_alignments,
+            "transition_vectors": transition_vectors,
+            "means": means,
+        }
+
+        return outputs
+
+    def train_step(self, batch: dict, criterion: nn.Module):
+        text_input = batch["text_input"]
+        text_lengths = batch["text_lengths"]
+        mel_input = batch["mel_input"]
+        mel_lengths = batch["mel_lengths"]
+
+        outputs = self.forward(
+            text=text_input,
+            text_len=text_lengths,
+            mels=mel_input,
+            mel_len=mel_lengths,
+        )
+
+        loss_dict = criterion(outputs["log_probs"])
+
+        return outputs, loss_dict
+
+    def eval_step(self, batch: Dict, criterion: nn.Module):
+        return self.train_step(batch, criterion)
+
+    def _format_aux_input(self, aux_input: Dict, default_input_dict):
+        """Set missing fields to their default value.
+
+        Args:
+            aux_inputs (Dict): Dictionary containing the auxiliary inputs.
+        """
+        default_input_dict.update(
+            {
+                "sampling_temp": self.sampling_temp,
+                "max_sampling_time": self.max_sampling_time,
+                "duration_threshold": self.duration_threshold,
+            }
+        )
+        if aux_input:
+            return format_aux_input(aux_input, default_input_dict)
+        return None
+
+    @torch.no_grad()
+    def inference(
+        self,
+        text: torch.Tensor,
+        aux_input={"x_lengths": None, "sampling_temp": None, "max_sampling_time": None, "duration_threshold": None},
+    ):  # pylint: disable=dangerous-default-value
+        """Sampling from the model
+
+        Args:
+            text (torch.Tensor): :math:`[B, T_in]`
+            aux_inputs (_type_, optional): _description_. Defaults to None.
+
+        Returns:
+            outputs: Dictionary containing the following
+                - mel (torch.Tensor): :math:`[B, T_out, C]`
+                - hmm_outputs_len (torch.Tensor): :math:`[B]`
+                - state_travelled (List[List[int]]): List of lists containing the state travelled for each sample in the batch.
+                - input_parameters (list[torch.FloatTensor]): Input parameters to the neural HMM.
+                - output_parameters (list[torch.FloatTensor]): Output parameters to the neural HMM.
+        """
+        # TODO: inputs are
```

Comment:
@erogol After the end of the epoch, while synthesising the test sentences, the inputs here are on CPU despite the model training on GPU. I am confused as to why that could be the case.

Reply:
It uses ```BaseTTS.test_run()``` if you want to debug what is wrong. It should be passing tensors to GPU normally.

Reply:
I have a hunch that it could be because of `nn.Parameters(required_grad=False)` and somehow the Trainer is not putting this parameter on the GPU correctly. I am trying to replace `nn.Parameters` with `register_buffer` and see if the problem disappears.

Reply:
yeap! That was it!